aslam_cv_common/include/aslam/common/unique-id.h

```diff
@@ -1,35 +1,255 @@
+    typedef std::size_t value_type;                             \
+    value_type operator()(const argument_type& hash_id) const { \
+      return hash_id.hashToSizeT();                             \
+    }                                                           \
+  };                                                            \
+  }                                                             \
+  extern void defineId##__FILE__##__LINE__(void)
+
+#define ALIAS_UNIQUE_ID(BaseIdType, AliasedIdType)  \
+  typedef BaseIdType AliasedIdType;                       \
+  typedef BaseIdType##List AliasedIdType##List;           \
+  typedef BaseIdType##Set AliasedIdType##Set
+
+template <typename ContainerType>
+inline std::string printIdContainer(const ContainerType& container) {
+  std::stringstream ss;
+  ss << "[ ";
+  for (const typename ContainerType::value_type& element : container) {
+    ss << element << " ";
+  }
+  ss << "]";
+  return ss.str();
+}
 
 namespace aslam {
-#define ASLAM_UNIQUE_ID(IdType)                   \
-  class IdType : public HashId {                  \
-   public:                                        \
-    inline static IdType Random() {               \
-      IdType generated;                           \
-      generated.randomize();                      \
-      return generated;                           \
-    }                                             \
-    IdType() = default;                           \
-  };                                              \
-  typedef std::vector<IdType> IdType##List;       \
-  typedef std::unordered_set<IdType> IdType##Set
-
-ASLAM_UNIQUE_ID(FrameId);
-ASLAM_UNIQUE_ID(NFramesId);
-ASLAM_UNIQUE_ID(CameraId);
-ASLAM_UNIQUE_ID(NCameraId);
+template <typename IdType>
+void generateId(IdType* id) {
+  CHECK_NOTNULL(id);
+  uint64_t hash[2];
+  internal::generateUnique128BitHash(hash);
+  id->fromUint64(hash);
+}
+
+template <typename IdType>
+IdType createRandomId() {
+  IdType id;
+  generateId(&id);
+  return id;
+}
+
+template <typename IdType>
+void generateIdFromInt(unsigned int idx, IdType* id) {
+  CHECK_NOTNULL(id);
+  std::stringstream ss;
+  ss << std::setfill('0') << std::setw(32) << idx;
+  id->fromHexString(ss.str());
+}
+
+// For internal use only.
+class Id : public HashId {
+ public:
+  Id() = default;
+  explicit inline Id(const aslam::proto::Id& id_field) {
+    deserialize(id_field);
+  }
+  inline void deserialize(const aslam::proto::Id& id_field) {
+    CHECK_EQ(id_field.uint_size(), 2);
+    fromUint64(id_field.uint().data());
+  }
+  inline void serialize(aslam::proto::Id* id_field) const {
+    CHECK_NOTNULL(id_field)->clear_uint();
+
+    google::protobuf::RepeatedField<google::protobuf::uint64>* uint_field =
+        id_field->mutable_uint();
+    uint_field->Add();
+    uint_field->Add();
+    toUint64(uint_field->mutable_data());
+  }
+
+  inline void fromHashId(const aslam::HashId& id) {
+    static_cast<aslam::HashId&>(*this) = id;
```

Comment:
Not completely sure here since it's not my code. But `Id` inherits from `HashId` and has no member variables, but only defines more interfaces on top. Therefore it's a way to initialize the inherited class object using a base class object. (No idea why this logic was done as such and why it's needed.)

Reply:
also, no clue, was like that before :P

Reply:
it seems like that it tries to call a specific copy assignment here?